File: go/bob/bob.go
// This is a "stub" file.  It's a little start on your solution.
// It's not a complete solution though; you have to write some code.

// Package bob should have a package comment that summarizes what it's about.
// https://golang.org/doc/effective_go.html#commentary
package bob

import (
	"regexp"
	"strings"
)

// Hey should have a comment documenting it.
func Hey(remark string) string {
	switch {
	case questionAndShouting(remark):
		return "Calm down, I know what I'm doing!"
	case isQuestion(remark):
		return "Sure"
	case isSilent(remark):
		return "Fine. Be that way!"
	case isShouting(remark):
		return "Whoa, chill out!"
	default:
		return "Whatever."
	}
}

func isQuestion(str string) bool {
	return strings.HasSuffix(sanitizeString(str), "?")
}

func questionAndShouting(str string) bool {
	return isQuestion(str) && isShouting(str) && !numbersOnly(str)
}

func isShouting(str string) bool {
	return strings.ToUpper(str) == str && !numbersOnly(str)
}

func isSilent(str string) bool {
	return sanitizeString(str) == ""
}

func sanitizeString(str string) string {
	return strings.Trim(str, " ")
}

func numbersOnly(str string) bool {
	newStr := prettifyString(str)
	return emptyString(sanitizeString(newStr))
}

func emptyString(str string) bool {
	return len(str) == 0
}

func prettifyString(str string) string {
	regex := regexp.MustCompile(`([0-9?:),])+`)
	return regex.ReplaceAllString(str, "")
}

File: elixir/bob/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      question_and_shouting?(input) -> "Calm down, I know what I'm doing!"
      question?(input) -> "Sure."
      silence?(input) -> "Fine. Be that way!"
      shouting?(input) -> "Whoa, chill out!"
      true -> "Whatever."
    end
  end

  def shouting?(input) do
    input == String.upcase(input) && !numbers_only?(input)
  end

  def question?(input) do
    String.trim(input) |> String.ends_with?("?")
  end

  def silence?(input) do
    String.trim(input) == ""
  end

  def question_and_shouting?(input) do
    question?(input) && shouting?(input) && !numbers_only?(input)
  end

  def numbers_only?(input) do
    prettyify_string(input) |>
    String.trim() |>
    empty_string?
  end

  def empty_string?(input) do
    String.length(input) == 0
  end

  def prettyify_string(input) do
    String.replace(input, ~r/([0-9?:),])+/, "")
  end
end
